### PR TITLE
syncer(dm): enable location recorder

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #4287

### What is changed and how it works?

don't let syncer maintain currentLocations, etc. Read them from streamController

#### Tests 

 - Unit test

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Comment by lance6716** [comment]
CI passed, will refine codes

https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/9096/pipeline

**Comment by lance6716** [comment]
> CI passed, will refine codes
> 
> https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/9096/pipeline

/cc @GMHDBJD @D3Hunter @buchuitoudegou PTAL

File: dm/syncer/binlogstream/streamer_controller.go [hunk 1]
```diff
@@ -333,35 +354,41 @@ func (c *StreamerController) GetEvent(tctx *tcontext.Context) (*replication.Binl
...
-	return event, suffix, op, nil
+	if isDataEvent(event) {
+		c.locations.curStartLocation = c.locations.curEndLocation
+		c.locations.curEndLocation = c.upstream.curEndLocation
+		c.locations.txnEndLocation = c.upstream.txnEndLocation
+		return event, op, nil
+	}
+
+	c.locations.curStartLocation = c.locations.curEndLocation
+	c.locations.curEndLocation.CopyWithoutSuffixFrom(c.upstream.curEndLocation)
```
Comment (D3Hunter) [inline dm/syncer/binlogstream/streamer_controller.go:384]: when will loc of upstream has suffix?

**Review by D3Hunter — COMMENTED** [review]
22/23 reviewed

File: dm/syncer/binlogstream/streamer_controller.go (hunk 1 above)
Comment (lance6716) [reply dm/syncer/binlogstream/streamer_controller.go:384]: upstream locations always has suffix 0, only injected events has non-zero suffix. When we reset to the middle of injected events whose suffix is not 0, upstream locations will overwrite the suffix without `CopyWithoutSuffixFrom`

File: dm/syncer/syncer.go [hunk 2]
```diff
@@ -1162,10 +1163,11 @@ func (s *Syncer) resetShardingGroup(table *filter.Table) {
...
 // we flush checkpoints in four cases:
-//   1. DDL executed
-//   2. pausing / stopping the sync (driven by `s.flushJobs`)
-//   3. IsFreshTask return true
-//   4. Heartbeat event received
+//  1. DDL executed
+//  2. pausing / stopping the sync (driven by `s.flushJobs`)
+//  3. IsFreshTask return true
+//  4. Heartbeat event received
+//
```
Comment (buchuitoudegou) [inline dm/syncer/syncer.go:1170]: Will XID Event trigger checkpoint flush?

**Review by buchuitoudegou — COMMENTED** [review]
Will review the test later😵‍💫

**Review by D3Hunter — APPROVED** [review]
lgtm

File: dm/syncer/syncer.go (hunk 2 above)
Comment (lance6716) [reply dm/syncer/syncer.go:1170]: XID only saves current end location to checkpoint in memory. flushing is done by other trigger

**Review by buchuitoudegou — APPROVED** [review]
lgtm

File: dm/syncer/binlogstream/streamer_controller.go [hunk 3]
```diff
@@ -277,41 +292,47 @@ func (c *StreamerController) resetReplicationSyncer(tctx *tcontext.Context, loca
...
 //     if we inject events [DDL1, DDL2] at (start) position 900, where start position
 //     900 has Insert1 event whose LogPos (end position) is 1000, we should return
 //     to caller like
-//   - DDL1, start position 900 LogPos 900, suffix 1
-//   - DDL2, start position 900 LogPos 900, suffix 2
-//   - Insert1, start position 900 LogPos 1000, suffix 0
+//
+//     1. DDL1, start (900, suffix 0) end (900, suffix 1)
+//     2. DDL2, start (900, suffix 1) end (900, suffix 2)
+//     3. Insert1, start (900, suffix 2) end (1000, suffix 0)
```
Comment (lyzx2001) [inline dm/syncer/binlogstream/streamer_controller.go:319]: Maybe can clarify more about the meaning of (900, suffix 0) and the other similar expressions

**Pull request merged**